### DISCUSSION TITLE
Loading Google Fonts dynamically

### DISCUSSION THREAD

**Discussion by emilienbidet** [vercel/next.js#51358] [Help]
### Summary

We have a multi tenant platform using the same code base.
Our clients are using different fonts so we are seeking a way to load the fonts (hopefully using next/font) dynamically per site.
We are investigating for a way to import Google Fonts dynamically.

Does any one have any option using next/font?
If no, I will use <link rel...> to import my fonts.

### Additional information

```js
import {
  ABeeZee,
  Abel,
  ... // importing all fonts
  Zilla_Slab,
  Zilla_Slab_Highlight,
} from "next/font/google";

const fonts = {
  ABeeZee,
  Abel,
  ... // all fonts
  Zilla_Slab,
  Zilla_Slab_Highlight,
};

export const getFont = (fontName: string) => {
  // @ts-ignore
  return fonts[fontName as keyof typeof fonts];
};

import "./globals.css";

...

export default async function RootLayout({
  children,
  params: { domain },
}: {
  children: React.ReactNode;
  params: {
    domain: string;
  };
}) {
  const site = await getSite(domain);

  // @ts-ignore
  const headingFont = getFont(site.font_heading)({
    variable: "--font-heading",
  });

  // @ts-ignore
  const bodyFont = getFont(site.font_body)({
    variable: "--font-body",
  });

  return (
    <html lang="fr" className={`${headingFont.variable} ${bodyFont.variable}`}>
      <body>
        {children}
      </body>
    </html>
  );
}


But I keep having this issue :

  x Font loaders must be called and assigned to a const in the module scope
      ,-[.../layout.tsx:3037:1]
 3037 |   Zeyada,
 3038 |   Zhi_Mang_Xing,
 3039 |   Zilla_Slab,
 3040 |   Zilla_Slab_Highlight,
      :   ^^^^^^^^^^^^^^^^^^^^
 3041 | };
 3042 | 
 3043 | export const getFont = (fontName: string) => {
      `----
```

### Related

I found Discord discussions on the NextJS Server about the subject.

- [Discord 1](https://discord.com/channels/752553802359505017/1064768118011998238)
- [Discord 2](https://discord.com/channels/752553802359505017/1089827381075791883)

**Comment by gerardnico** [comment]
Did you try [dynamic import](https://datacadamia.com/web/javascript/module/dynamic_import)?


```javascript
import('next/font/google').then( ({ Roboto }) => {
   ...
});
```

**Reply by emilienbidet** [reply]
I just tried. I can't succeed to make it works with dynamic import.

**Comment by emilienbidet** [comment]
@hanneslund Do you have any idea on how to achieve this behaviour ?

**Comment by emilienbidet** [comment]
I found a work around that works but fonts aren't optimised and self hosted.

```tsx
export default async function RootLayout({
  children,
  params: { domain },
}: {
  children: React.ReactNode;
  params: {
    domain: string;
  };
}) {
  const site = await getSite(domain);

  if (!site) {
    notFound();
  }

  // replace space with '+'
  const headingFont = site.font_heading.replace(" ", "+");
  const bodyFont = site.font_body.replace(" ", "+");

  // Create font URL for Google Fonts
  const fontUrl = `https://fonts.googleapis.com/css2?family=${headingFont}&family=${bodyFont}&display=swap`;

  return (
    <html lang="fr">
      <head className="">
        <link rel="preconnect" href="https://fonts.googleapis.com" />
        <link
          rel="preconnect"
          href="https://fonts.gstatic.com"
          crossOrigin="anonymous"
        />
        <link href={fontUrl} rel="stylesheet" />
        <style>
          {`
            :root {
              --font-heading: ${site.font_heading};
              --font-body: ${site.font_body};
            }
          `}
        </style>
      </head>
      <body>
        {children}
      </body>
    </html>
  );
}

```

**Comment by artificeryy** [comment]
Hey @emilienbidet, I'm also interested in whether it's possible to dynamically load in Google fonts based on user input. So far, I've also arrived at the same solution of getting the fonts from the Google Fonts API.

**Reply by msalahz** [reply]
Any updates here?

**Reply by mmswi** [reply]
Same here, we need this too!

**Reply by rosswilson2306** [reply]
Likewise, would be interested in a solution for this. We've been using @emilienbidet's solution but it's affecting our LCP score.

**Reply by devon2018** [reply]
Has anyone found a solution for this?

**Reply by mmswi** [reply]
getting the fonts from the api. also when using just the inter next font, it was different than the one hosted on Google. so will stick to the hosted fonts for now.

**Comment by emilienbidet** [comment]
I think that the easy solution would be to download and to store all the fonts **you need** under the `public/fonts` folder.

And then build your font face dynamically.

You can help maybe be use the [FontSource API](https://fontsource.org/docs/api/introduction) it will speed up your implementation.

I'm not working on this project right now, I can't work on a cleaner method.
